### PR TITLE
Converted form Element in RegisterPage and LoginPage to MUI components

### LINKED ISSUE #65: Migrate form to MUI

### PR DESCRIPTION
**Pull request by dyazdani**
Closes #65

RegisterPage:
Screen Shot 2023-11-14 at 12 26 59

constraint validation:
Screen Shot 2023-11-14 at 12 44 05
Screen Shot 2023-11-14 at 12 27 17
Screen Shot 2023-11-14 at 12 27 26
Screen Shot 2023-11-14 at 12 28 05
Screen Shot 2023-11-14 at 12 28 33

navigates to home page upon successful registration:
Screen Shot 2023-11-14 at 12 29 06

LoginPage:
Screen Shot 2023-11-14 at 12 39 32

constraint validation:
Screen Shot 2023-11-14 at 12 46 57
Screen Shot 2023-11-14 at 12 47 16

navigates to home page upon successful login:
Screen Shot 2023-11-14 at 12 40 29 copy

### CODE REVIEW

File: client/components/LoginPage.tsx [hunk 1]
```diff
@@ -1,44 +1,93 @@
 import React, { useState } from "react";
+import { useNavigate } from "react-router-dom";
+import { 
+  TextField, 
+  Box, 
+  Typography, 
+  InputAdornment,
+  IconButton,
+  Button
+} from "@mui/material";
```
Comment (paloobi) [inline client/components/LoginPage.tsx:10]: This imports the entire mui component library, when you only need a few specific components. Depending on dev setup, this can be very very bad for bundle sizes (I'm not 100% sure if Vite handles this through what is called "tree shaking" but I have seen some evidence from people that this has caused issues for them with Vite builds, so I'm suspecting it will not protect from this). This is why the MUI documentation examples mostly do not import components this way. I recommend instead importing each component directly from its module

```suggestion
import TextField from "@mui/material/TextField";
import Box from "@mui/material/Box";
import Typography from "@mui/material/Typography";
import InputAdornment from "@mui/material/InputAdornment";
import IconButton from "@mui/material/IconButton";
import Button from "@mui/material/Button";
```

File: client/components/LoginPage.tsx [hunk 2]
```diff
@@ -1,44 +1,93 @@
...
+import {
+  Visibility,
+  VisibilityOff
+} from "@mui/icons-material";
```
Comment (paloobi) [inline client/components/LoginPage.tsx:14]: Same note as above, except the documentation suggests naming these imports with the Icon prefix to avoid naming collisions.

```suggestion
import VisibilityIcon from "@mui/icons-material/Visibility";
import VisibilityOffIcon from "@mui/icons-material/VisibilityOff";
```

File: client/components/LoginPage.tsx [hunk 3]
```diff
@@ -1,44 +1,93 @@
...
 
+  const navigate = useNavigate();
+
+
   const [email, setEmail] = useState("");
   const [password, setPassword] = useState("");
+  const [showPassword, setShowPassword] = useState(false);
+  const [showConfirmPassword, setShowConfirmPassword] = useState(false);
+
+  const handleClickShowPassword = () => setShowPassword((show) => !show);
```
Comment (paloobi) [inline client/components/LoginPage.tsx:29]: I would recommend just putting this 1-line function directly as the onClick rather than defining it here - it will be better for readability of your code. It is not necessary to always define handlers as a separate function, and honestly not preferable imo if it's a single line like this as it makes it harder when looking at the JSX below to tell what it does

File: client/components/LoginPage.tsx [hunk 4]
```diff
@@ -1,44 +1,93 @@
...
+  const handleMouseDownPassword = (event: React.MouseEvent<HTMLButtonElement>) => {
+    event.preventDefault();
+  };
```
Comment (paloobi) [inline client/components/LoginPage.tsx:32]: What does this do and why do we need it? Maybe add a comment to explain why.

Also might recommend putting this in-line with the onMouseDown prop below instead for readability.

Do you actually need to type the event? I thought that TS could infer, but maybe not

File: client/components/LoginPage.tsx [hunk 5]
```diff
@@ -1,44 +1,93 @@
...
-        />
-
-        <button disabled={isLoading} type="submit">
-          Login
-        </button>
-      </form>
+    <Box 
+      component="form"
+      onSubmit={handleSubmit}
+      autoComplete="off"
```
Comment (paloobi) [inline client/components/LoginPage.tsx:44]: What was your reasoning for turning off this default browser behavior?

File: client/components/LoginPage.tsx [hunk 6]
```diff
@@ -1,44 +1,93 @@
...
+    >
+      <Typography component="h1">Login</Typography>
```
Comment (paloobi) [inline client/components/LoginPage.tsx:46]: This is probably an h2, not an h1 - the h1 would be the name of the app, and there should generally only be one h1 on the page

File: client/components/LoginPage.tsx [hunk 7]
```diff
@@ -1,44 +1,93 @@
...
+      {isError && <Typography 
+        paragraph={true}
+      >
+        {/*TODO: Change this to an alert*/}Oops, there was an error creating an account. Try again?
+      </Typography>}
+
+      <TextField
+        id="email"
```
Comment (paloobi) [inline client/components/LoginPage.tsx:54]: Does this need an ID for some reason? Unless you are using it somewhere I recommend removing

Comment (paloobi) [inline client/components/LoginPage.tsx:54]:
```suggestion
```

File: client/components/LoginPage.tsx [hunk 8]
```diff
@@ -1,44 +1,93 @@
...
+        label="Email"
+        type="email"
+        required
+        onChange={(e) => setEmail(e.target.value)}
+        value={email}
+      />
+
+      <TextField
+        id="password"
```
Comment (paloobi) [inline client/components/LoginPage.tsx:63]: Same - delete unless you absolutely need it an are already using it
```suggestion
```

File: client/components/LoginPage.tsx [hunk 9]
```diff
@@ -1,44 +1,93 @@
...
+                aria-label="toggle password visibility"
+                onClick={handleClickShowPassword}
+                onMouseDown={handleMouseDownPassword}
+                edge="end"
+              >
+                {showPassword ? <VisibilityOff /> : <Visibility/>}
+              </IconButton>
+            </InputAdornment>
+        }}
+        onChange={(e) => setPassword(e.target.value)}
```
Comment (paloobi) [inline client/components/LoginPage.tsx:80]: same nit: no need for extra parens
```suggestion
        onChange={e => setPassword(e.target.value)}
```

File: client/components/RegisterPage.tsx [hunk 10]
```diff
@@ -1,64 +1,133 @@
 import React, { useState } from "react";
+import { useNavigate } from "react-router-dom";
+import { 
+  TextField, 
+  Box, 
+  Typography, 
+  InputAdornment,
+  IconButton,
+  Button
+} from "@mui/material";
```
Comment (paloobi) [inline client/components/RegisterPage.tsx:10]: Same note as above - don't import the whole package
```suggestion
import TextField from "@mui/material/TextField";
import Box from "@mui/material/Box";
import Typography from "@mui/material/Typography";
import InputAdornment from "@mui/material/InputAdornment";
import IconButton  from "@mui/material/IconButton";
import Button  from "@mui/material/Button";
```

File: client/components/RegisterPage.tsx [hunk 11]
```diff
@@ -1,64 +1,133 @@
...
+import {
+  Visibility,
+  VisibilityOff
+} from "@mui/icons-material";
```
Comment (paloobi) [inline client/components/RegisterPage.tsx:14]: same as above
```suggestion
import VisibilityIcon from "@mui/icons-material/Visibility";
import VisibilityOffIcon "@mui/icons-material/VisibilityOff";
```

File: client/components/RegisterPage.tsx [hunk 12]
```diff
@@ -1,64 +1,133 @@
...
+  const navigate = useNavigate();
+
   const [email, setEmail] = useState("");
   const [username, setUsername] = useState("");
   const [password, setPassword] = useState("");
   const [confirmPassword, setConfirmPassword] = useState("");
+  const [showPassword, setShowPassword] = useState(false);
+  const [showConfirmPassword, setShowConfirmPassword] = useState(false);
+
+  const handleClickShowPassword = () => setShowPassword((show) => !show);
```
Comment (paloobi) [inline client/components/RegisterPage.tsx:29]: same note as Login - recommend defining the function inline

File: client/components/RegisterPage.tsx [hunk 13]
```diff
@@ -1,64 +1,133 @@
...
+  const handleMouseDownPassword = (event: React.MouseEvent<HTMLButtonElement>) => {
+    event.preventDefault();
+  };
```
Comment (paloobi) [inline client/components/RegisterPage.tsx:32]: Same note as above - recommend defining inline, and not explicitly typing the Event unless TS gets mad

File: client/components/RegisterPage.tsx [hunk 14]
```diff
@@ -1,64 +1,133 @@
...
+
+  const handleClickShowConfirmPassword = () => setShowConfirmPassword((show) => !show);
```
Comment (paloobi) [inline client/components/RegisterPage.tsx:34]: Same - recommend defining inline in the props

File: client/components/RegisterPage.tsx [hunk 15]
```diff
@@ -1,64 +1,133 @@
...
+  const handleMouseDownConfirmPassword = (event: React.MouseEvent<HTMLButtonElement>) => {
+    event.preventDefault();
+  };
```
Comment (paloobi) [inline client/components/RegisterPage.tsx:37]: same as above - inline + double-check if Event needs to be explicitly typed

File: client/components/RegisterPage.tsx [hunk 16]
```diff
@@ -1,64 +1,133 @@
...
+  };  
 
   const handleSubmit = (e: React.FormEvent<HTMLFormElement>) => {
     e.preventDefault();
-    register({ email, username, password });
+    if (password === confirmPassword) {
+      register({ email, username, password });
+      navigate("/")
+    } else {
+      alert("Password confirmation does not match");
```
Comment (paloobi) [inline client/components/RegisterPage.tsx:45]: This should be a React error that is displayed on the UI - if you want to come back for it later when you add more detailed form validation, please leave a TODO

File: client/components/RegisterPage.tsx [hunk 17]
```diff
@@ -1,64 +1,133 @@
...
   };
 
   return (
-      <form onSubmit={handleSubmit}>
-        <h1>Register</h1>
-        {isError && <p>Oops, there was an error creating an account. Try again?</p>}
+      <Box 
+        component="form"
+        onSubmit={handleSubmit}
+        autoComplete="off"
```
Comment (paloobi) [inline client/components/RegisterPage.tsx:53]: same note as with login - confirming that you want to turn off this default browser behavior and why

File: client/components/RegisterPage.tsx [hunk 18]
```diff
@@ -1,64 +1,133 @@
...
+          paragraph={true}
+        >
+          {/*TODO: Change this to an alert*/}Oops, there was an error creating an account. Try again?
+        </Typography>}
 
-        <label htmlFor="email">Email</label>
-        <input
-          type="email"
+        <TextField
           id="email"
```
Comment (paloobi) [inline client/components/RegisterPage.tsx:63]: Delete unless used somewhere

File: client/components/RegisterPage.tsx [hunk 19]
```diff
@@ -1,64 +1,133 @@
...
-          name="email"
+          label="Email"
+          type="email"
+          required
           onChange={(e) => setEmail(e.target.value)}
```
Comment (paloobi) [inline client/components/RegisterPage.tsx:67]: nit
```suggestion
          onChange={e => setEmail(e.target.value)}
```

File: client/components/RegisterPage.tsx [hunk 20]
```diff
@@ -1,64 +1,133 @@
...
           value={email}
         />
 
-        <label htmlFor="username">Username</label>
-        <input
-          type="text"
+        <TextField
           id="username"
```
Comment (paloobi) [inline client/components/RegisterPage.tsx:72]: delete if not used
```suggestion
```

File: client/components/RegisterPage.tsx [hunk 21]
```diff
@@ -1,64 +1,133 @@
...
-          name="username"
+          label="Username"
+          type="text"
+          required
           onChange={(e) => setUsername(e.target.value)}
```
Comment (paloobi) [inline client/components/RegisterPage.tsx:76]: nit
```suggestion
          onChange={e => setUsername(e.target.value)}
```

File: client/components/RegisterPage.tsx [hunk 22]
```diff
@@ -1,64 +1,133 @@
...
           value={username}
         />
 
-        <label htmlFor="password">Password</label>
-        <input
-          type="password"
+        <TextField
           id="password"
```
Comment (paloobi) [inline client/components/RegisterPage.tsx:81]: delete if unused
```suggestion
```

File: client/components/RegisterPage.tsx [hunk 23]
```diff
@@ -1,64 +1,133 @@
...
+          }}
           onChange={(e) => setPassword(e.target.value)}
           value={password}
         />
 
-        <label htmlFor="confirmPassword">Confirm Password</label>
-        <input
-          type="password"
+        <TextField
           id="confirmPassword"
```
Comment (paloobi) [inline client/components/RegisterPage.tsx:103]: delete if unused
```suggestion
```

File: public/bundle.js
Comment (paloobi) [inline public/bundle.js:1]: hmm, not something to address in this PR, but I don't know if it's a good idea for the built bundle to be stored in git... usually built/generated files are not stored in git, they are generated instead on device

File: client/components/RegisterPage.tsx [hunk 24]
```diff
@@ -1,64 +1,133 @@
...
   return (
-      <form onSubmit={handleSubmit}>
-        <h1>Register</h1>
-        {isError && <p>Oops, there was an error creating an account. Try again?</p>}
+      <Box 
+        component="form"
+        onSubmit={handleSubmit}
+        autoComplete="off"
+      >
+        <Typography component="h1">Register</Typography>
```
Comment (paloobi) [inline client/components/RegisterPage.tsx:55]: Probably should be an h2

**Review by paloobi — CHANGES_REQUESTED** [review]
RC for:

* don't import the entire MUI library when using individual components / libraries
* h1 -> h2
* delete unused IDs in forms

Also wanted to check in - is your plan to do that in a separate PR from converting to MUI? If so I'm ok if so, I just noticed the form layout and alignment doesn't look finished

File: client/components/LoginPage.tsx (hunk 1 above)
Comment (dyazdani) [reply client/components/LoginPage.tsx:10]: Got it. I actually did not know the difference. Definitely don't want to import more than is needed

File: client/components/LoginPage.tsx (hunk 3 above)
Comment (dyazdani) [reply client/components/LoginPage.tsx:24]: Got it. I think I got this from the MUI docs.

File: client/components/LoginPage.tsx (hunk 4 above)
Comment (dyazdani) [reply client/components/LoginPage.tsx:27]: TS didn't infer correctly when I was doing it. This is cut and pasted from the MUI docs. I think it's included so that the icon does not change with just a mouse down event, but rather a full click.

Comment (dyazdani) [reply client/components/LoginPage.tsx:27]: I can put it inline in onMouseDown

File: client/components/LoginPage.tsx (hunk 5 above)
Comment (dyazdani) [reply client/components/LoginPage.tsx:39]: MDN says it can be used as a security measure. It's included in all of the MUI examples.

File: client/components/RegisterPage.tsx (hunk 16 above)
Comment (dyazdani) [reply client/components/RegisterPage.tsx:32]: Added TODO

File: public/bundle.js
Comment (dyazdani) [reply public/bundle.js:1]: does `bundle.js` need to be added to the .gitignore file?

**Comment by dyazdani** [comment]
Yes, this PR was just for converting to MUI without any formatting.

Regarding formatting (spacing, margins, etc.): how do you keep formatting consistent throughout the app when members of the team are working on different features async? Is everything defined globally in the CSS? Surely some pages would have to have at least slightly different formatting. Do you just communicate with the team and maybe make a doc stating how the app is being designed? E.g., if the app is using grids or flexbox.

File: public/bundle.js
Comment (paloobi) [reply public/bundle.js:1]: Yeah probably a separate PR should add it to gitignore and also delete it from the repo. maybe worth filing an Issue for that

**Comment by paloobi** [comment]
For styling, the idea is the start by styling a component (if there is no mock you may want to make a quick wireframe of what the register and login pages would look like).

Remember - no CSS needed. Layout will be handled with MUI, likely using Box. Agree on the outcome, not the implementation (flex vs. grid) - that's what mocks help with. For things like margin/padding - keep it consistent, and review each others code to ensure it is consistent. If you find you are disagreeing, note it in the PR and come to an agreement (eg. I make a form with 4px margin between all the elements, you make a form with 8px margin - I review your code, and I notice you used 8px. We have a conversation and agree on whether to use 4 or 8 - if we decide 8, I change my styles, if you decide 4, then the PR is updated)

Basically - there isn't a simple way to keep it consistent other than to agree or notice when things are not consistent. I can help notice design discrepancies during a design review at the end.

tldr; if you are working with a designer, they would provide you a spec for all margin/padding/spacing and you would just follow the spec (unless you notice an issue/inconsistency/impossible style and need to ask for a revision on the design). No designer means you'll be making the spec as you go.

File: client/components/LoginPage.tsx [hunk 25]
```diff
@@ -1,44 +1,81 @@
...
+        type={showPassword ? "text" : "password"}
+        required
+        InputProps={{
+          endAdornment:
+            <InputAdornment position="end">
+              <IconButton
+                aria-label="toggle password visibility"
+                onClick={() => setShowPassword((show) => !show)}
+                onMouseDown={(e: React.MouseEvent<HTMLButtonElement>) => e.preventDefault()}
+              edge="end"
```
Comment (paloobi) [inline client/components/LoginPage.tsx:62]: nit: indentation
```suggestion
                edge="end"
```

File: client/components/RegisterPage.tsx (hunk 16 above)
Comment (paloobi) [reply client/components/RegisterPage.tsx:32]: Note: if you have a GitHub issue created for this already, you can add it to the TODO, eg `TODO #32 Change this to a React error that is displayed in the UI`

**Review by paloobi — APPROVED** [review]
Looks good to me!
